net/gcoap: Refine Observe re-registration

kb2ma opened the pull request:
### Contribution description
This PR builds on the fix in #9255 to handle another case, and to clarify the logic for when the observer and resource may be set for an Observe registration. Specifically:

- Fix re-register when using the same token (still a problem in #9235).
- Disallow re-register using a token already used for another registration.
- Only set observer and resource on initial registration.

### Issues/PRs references
Alternative to #9255.

Fixes #9235.

miri64 commented:
> * Disallow re-register using a token already used for another registration.

Can you elaborate on this? Because I don't understand right now how this is an error or why this should be disallowed. Isn't the token client generated? So if another client has the same token, this should still work.

kb2ma commented:
See [example 1](https://github.com/RIOT-OS/RIOT/pull/9255#issuecomment-394091289) and [example 2](https://github.com/RIOT-OS/RIOT/pull/9255#issuecomment-394091343) for corner cases with #9255.

Changing the token really impacts the identity of the registration itself. The pre-existing code that triggered issue #9180 wasn't designed for this kind of re-registration. My goal is to be just invasive enough to solve the problems.

BTW, the code here still is WIP, and I'm working to make it a little easier to understand. It should be ready for review in a day or so.

miri64 commented:
I still don't understand the troubling nature of these two examples. For me they look like instances that may happen, either because the client is faulty or because the token *happens* to be the same again (which with the non-infinite set of numbers the token is choosable from may always happen). Due to the tuple `(remote_addr, remote_port, token, path)` the registrations should still be distinguishable. It should be the server's responsibility to deal with this kind of edge cases and not punish the client for it.

Or is this PR making this possible in the first case?

kb2ma commented:
For example 1, memo is not NULL and observer is NULL. This means the memo's observer attribute will be set to NULL in #9255, which results in an invalid registration.

#9261 will not accept the registration in example 1, for two reasons. It will mean that two observers have registered for the same resource. For simplicity, gcoap adds the constraint that only a single observer may register for a resource. Also, it would mean that aaaa::1 has two registrations with the same token. (See below.)

For example 2, two registrations for a single endpoint+token is ambiguous. Suppose the client receives a notification with token 2. Does the notification refer to resource /a or to resource /b?

I think I'm finished with coding, so I pushed to GitHub. Still WIP because I had to update my functional test framework to allow specification of a particular token when registering. That's working now, so I'm writing the tests.

miri64 commented:
Ok. I somehow forgot that the token is used to identify the registration ^^"

kb2ma commented:
No worries, the collaboration and discussion definitely have improved the implementation. :-)

I just removed the WIP label and would appreciate your review. I added several [functional tests](https://github.com/kb2ma/RIOT/wiki/Observe-Tests) ([source](https://github.com/kb2ma/gcoap-test/blob/master/expect/observe_test.py)) for re-registration, including the edge cases we discussed.

miri64 commented:
Murdock's happy, so go!
